homeassistant/components/binary_sensor/trend.py

```diff
@@ -106,32 +135,51 @@ def device_class(self):
         return self._device_class
 
     @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        return {
+            ATTR_ENTITY_ID: self._entity_id,
+            ATTR_FRIENDLY_NAME: self._name,
+            ATTR_INVERT: self._invert,
+            ATTR_GRADIENT: self._gradient,
+            ATTR_MIN_GRADIENT: self._min_gradient,
+            ATTR_SAMPLE_DURATION: self._sample_duration,
+            ATTR_SAMPLE_COUNT: len(self.samples),
+        }
+
+    @property
     def should_poll(self):
         """No polling needed."""
         return False
 
     @asyncio.coroutine
     def async_update(self):
         """Get the latest data and update the states."""
-        if self.from_state is None or self.to_state is None:
-            return
-        if (self.from_state.state == STATE_UNKNOWN or
-                self.to_state.state == STATE_UNKNOWN):
+        import math
+        import numpy as np
+
+        # Remove outdated samples
+        if self._sample_duration == 0:
+            self.samples = self.samples[-2:]
+        else:
+            now = datetime.now(timezone.utc).timestamp()
+            cutoff = now - self._sample_duration
+            self.samples = [(t, v) for t, v in self.samples if t >= cutoff]
```

Comment:
That is to slow. Remove that from deque until the first is inside our time area and stop that. Also don't copy the data every time inside memory.